### PR TITLE
[NTOS:KE/x64] Implement KeDisconnectInterrupt for amd64

### PR DESCRIPTION
**Pull request by TheNNX**
## Purpose

Implement `KeDisconnectInterrupt` for amd64.

## Proposed changes

- Implement `KeDisconnectInterrupt` for amd64
- Make `KeConnectInterrupt` lock the dispatcher lock

### CODE REVIEW

**Review by HBelusca — COMMENTED** [review]
This looks a bit similar to the x86 one (but without that PDISPATCH_INFO thing). I cannot fully judge the exactness of the proposed code, but it looks reasonable.

File: ntoskrnl/ke/amd64/interrupt.c
Comment (tkreuzer) [inline ntoskrnl/ke/amd64/interrupt.c:186]: I think it's more intuitive to compare the interrupt objects instead of the list entry, i.e. `ASSERT(Interrupt == VectorFirstInterrupt);` same for the `else if` below.

**Review by tkreuzer — CHANGES_REQUESTED** [review]

**Comment by TheNNX** [comment]
Thank you for the code review. I believe I have now implemented the changes requested.

File: ntoskrnl/ke/amd64/interrupt.c
Comment (binarymaster) [inline ntoskrnl/ke/amd64/interrupt.c:126]:
```suggestion
            goto Cleanup;
```

File: ntoskrnl/ke/amd64/interrupt.c
Comment (binarymaster) [inline ntoskrnl/ke/amd64/interrupt.c:139]:
```suggestion
            goto Cleanup;
```

File: ntoskrnl/ke/amd64/interrupt.c
Comment (binarymaster) [inline ntoskrnl/ke/amd64/interrupt.c:149]:
```suggestion
Cleanup:
```

**Review by binarymaster — COMMENTED** [review]
I think this label name would be more consistent with existing labels we have in the kernel.

File: ntoskrnl/ke/amd64/interrupt.c
Comment (tkreuzer) [inline ntoskrnl/ke/amd64/interrupt.c:217]: This is obsolete now.

Comment (TheNNX) [reply ntoskrnl/ke/amd64/interrupt.c:217]: Indeed, it is. Should I remove it? I'm asking because you've approved the changes already.

Comment (binarymaster) [reply ntoskrnl/ke/amd64/interrupt.c:217]: I think removing the dead code makes sense regardless of the approval.

Comment (TheNNX) [reply ntoskrnl/ke/amd64/interrupt.c]: Makes sense. I was a bit confused by the approval together with the comment, though.

Comment (binarymaster) [reply ntoskrnl/ke/amd64/interrupt.c]: Usually it means "here, you have my approval, but make sure to address this before merging." :slightly_smiling_face:

File: ntoskrnl/ke/amd64/interrupt.c
Comment (binarymaster) [inline ntoskrnl/ke/amd64/interrupt.c:216]:
```suggestion

```

File: ntoskrnl/ke/amd64/interrupt.c
Comment (binarymaster) [inline ntoskrnl/ke/amd64/interrupt.c:149]:
```suggestion

Cleanup:
```

**Review by binarymaster — COMMENTED** [review]
Few whitespace problems found (will fix them up myself via GitHub).

**Review by HBelusca — APPROVED** [review]

**Pull request merged**
